server/handlers.go

```diff
@@ -1142,22 +1155,28 @@ func (s *Server) handleRefreshToken(w http.ResponseWriter, r *http.Request, clie
 		return
 	}
 
-	newToken := &internal.RefreshToken{
-		RefreshId: refresh.ID,
-		Token:     storage.NewID(),
-	}
-	rawNewToken, err := internal.Marshal(newToken)
-	if err != nil {
-		s.logger.Errorf("failed to marshal refresh token: %v", err)
-		s.tokenErrHelper(w, errServerError, "", http.StatusInternalServerError)
-		return
+	newToken := token
+	if s.refreshTokenPolicy.RotationEnabled() {
+		newToken = &internal.RefreshToken{
+			RefreshId: refresh.ID,
+			Token:     storage.NewID(),
+		}
 	}
 
 	lastUsed := s.now()
 	updater := func(old storage.RefreshToken) (storage.RefreshToken, error) {
-		if old.Token != refresh.Token {
-			return old, errors.New("refresh token claimed twice")
+		if s.refreshTokenPolicy.RotationEnabled() {
+			if old.Token != refresh.Token {
+				if s.refreshTokenPolicy.AllowedToReuse(old.LastUsed) && old.ObsoleteToken == refresh.Token {
+					newToken.Token = old.Token
```

Comment:
I would expect old.Claims and old.LastUsed to be updated in this scenario as well

Reply:
Every request must be responded with the same refresh token and id token. Updating claims here will change that logic. We also should avoid calling the refresh method of connectors.

Need a second opinion here.

Reply:
I don't quite follow. refreshConn.Refresh() is called earlier in this method regardless of the update here. And what do you mean by:
> Every request must be responded with the same refresh token and id token

Reply:
The question is that maybe we don't need to call connector refresh for requests with an obsolete token? This is an open question to discuss.

Reply:
Added the clause to make dex skip refreshing claims using a connector on requests with an obsolete token.